core/src/main/java/hivemall/mf/CofactorModel.java

```diff
@@ -0,0 +1,715 @@
+            dst.setEntry(i, src[i]);
+        }
+    }
+
+    private static void copyData(double[] dst, RealVector src) throws HiveException {
+        checkCondition(dst.length == src.getDimension(), DIFFERENT_DIMS_ERR);
+        for (int i = 0; i < dst.length; i++) {
+            dst[i] = src.getEntry(i);
+        }
+    }
+
+    @VisibleForTesting
+    protected static double[][] calculateWTW(Map<String, double[]> weights, int numFactors, float constant) {
+        double[][] WTW = new double[numFactors][numFactors];
+        for (int f = 0; f < numFactors; f++) {
+            for (int ff = 0; ff < numFactors; ff++) {
+                double val = constant * dotFactorsAlongDims(weights, f, ff);
+                WTW[f][ff] = val;
+            }
+        }
+        return WTW;
+    }
+
+    @VisibleForTesting
+    protected static double[][] calculateWTWSubset(List<Feature> subset, Map<String, double[]> weights, int numFactors, float constant) {
+        // equivalent to `B_u.T.dot((c1 - c0) * B_u)` in cofacto.py
+        double[][] delta = new double[numFactors][numFactors];
+        for (int f = 0; f < numFactors; f++) {
+            for (int ff = 0; ff < numFactors; ff++) {
+                double val = constant * dotFactorsAlongDims(subset, weights, f, ff);
+                delta[f][ff] = val;
+            }
+        }
+        return delta;
+    }
+
+    private static double dotFactorsAlongDims(List<Feature> keys, Map<String, double[]> weights, int dim1, int dim2) {
+        double result = 0.d;
+        for (Feature f : keys) {
+            double[] vec = getFactorVector(f.getFeature(), weights);
+            result += vec[dim1] * vec[dim2];
+        }
+        return result;
+    }
+
+    private static double dotFactorsAlongDims(Map<String, double[]> weights, int dim1, int dim2) {
+        double result = 0.d;
+        for (double[] vec : weights.values()) {
+            result += vec[dim1] * vec[dim2];
+        }
+        return result;
+    }
+
+    @VisibleForTesting
+    protected static double[] calculateA(List<Feature> items, Map<String, double[]> weights, int numFactors, float constant) throws HiveException {
+        // Equivalent to: a = x_u.dot(c1 * B_u)
+        // x_u is a (1, i) matrix of all ones
+        // B_u is a (i, F) matrix
+        // What it does: sums factor n of each item in B_u
+//        clearArray(A);
+        double[] A = new double[numFactors];
+        for (Feature item : items) {
+            double y_ui = item.getValue(); // rating
+            addInPlace(A, getFactorVector(item.getFeature(), weights), y_ui);
+        }
+        for (int a = 0; a < A.length; a++) {
+            A[a] *= constant;
+        }
+        return A;
+    }
+
+    private static void clearArray(double[] a) {
+        for (int i = 0; i < a.length; i++) {
+            a[i] = 0.d;
+        }
+    }
+
+    public Double predict(String user, String item) {
+        if (!isTrainable(user, theta) || !isTrainable(item, beta)) {
```

Comment:
naming can be revised. `isTrainable` in predict? 

Rename or simply removing method. `!theta.contains(user) || beta.contains(item)`